src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessorWorkflow.java

```diff
@@ -226,6 +252,201 @@ private void updateOriginalQueryResults(final CombineScoresDto combineScoresDTO,
         }
     }
 
+    private int collapse(
+        List<CompoundTopDocs> queryTopDocs,
+        int indexOfFirstNonEmpty,
+        List<QuerySearchResult> querySearchResults,
+        Sort sort,
+        final boolean isFetchPhaseExecuted,
+        CombineScoresDto combineScoresDTO
+    ) {
+        boolean isKeyword = ((CollapseTopFieldDocs) queryTopDocs.get(indexOfFirstNonEmpty)
+            .getTopDocs()
+            .getFirst()).collapseValues[0] instanceof BytesRef;
+
+        if (isKeyword) {
+            return collapseOnKeyword(queryTopDocs, querySearchResults, sort, indexOfFirstNonEmpty, isFetchPhaseExecuted, combineScoresDTO);
+        } else {
+            return collapseOnNumeric(queryTopDocs, querySearchResults, sort, indexOfFirstNonEmpty, isFetchPhaseExecuted, combineScoresDTO);
+        }
+    }
+
+    private int collapseOnKeyword(
+        List<CompoundTopDocs> queryTopDocs,
+        List<QuerySearchResult> querySearchResults,
+        Sort sort,
+        int indexOfFirstNonEmpty,
+        final boolean isFetchPhaseExecuted,
+        CombineScoresDto combineScoresDTO
+    ) {
+        int totalScoreDocsCount = 0;
+        Map<BytesRef, FieldDoc> collapseValueToTopScoreDocMap = new HashMap<>();
+        Map<BytesRef, Integer> collapseValueToShardMap = new HashMap<>();
+        String collapseField = "";
+
+        HybridQueryFieldDocComparator comparator = new HybridQueryFieldDocComparator(
+            ((CollapseTopFieldDocs) queryTopDocs.get(indexOfFirstNonEmpty).getTopDocs().getFirst()).fields,
+            Comparator.comparing((ScoreDoc scoreDoc) -> scoreDoc.score)
+        );
+
+        for (int i = 0; i < querySearchResults.size(); i++) {
```

Comment:
can you rename i to signify what are we iterating on. I assume this is shardIndex